### PR TITLE
Add forward and train methods to spectralFFN()  

### PR DESCRIPTION
**Pull request by am2145**
## Description
This PR modifies the `spectralFFN()` class in `predictors.py` to add `forward()` and `train_step()` functions in line with the other FFN predictors. Predictions are normalized and the spectral activation function is used to enforce positivity, both in line with the spectral prediction approach in v1. 

## Questions
1) @hwpang and I have done initial testing using spectra.csv for the base case, and for training to behave as expected, the targets also need to be normalized. We can either normalize the targets during the dataloading process (which was done in v1), or normalize targets during the loss calculation and testing instead. Which is the preferred approach?

### CODE REVIEW

**Comment by davidegraff** [comment]
I would normalize the targets prior to training so we're not redoing the same work in each epoch. However, that opens up another can of worms in terms of implementation because ideally we don't want to have some bespoke normalization method for each type of task.

**Comment by am2145** [comment]
> I would normalize the targets prior to training so we're not redoing the same work in each epoch. However, that opens up another can of worms in terms of implementation because ideally we don't want to have some bespoke normalization method for each type of task.

Normalizing the targets prior to training makese sense and is consistent with v1, which is and advantage. However, I also see your point that we don't want to have too many normalizations floating around. From my reading of v1, it checks if the dataset type is spectral and then calls a normalizing function, but does not do this for other dataset types (at least as the default). Would a similar approach seem appropriate to you here where we only default to normalization for spectralFFN?

**Comment by davidegraff** [comment]
So to be clear, what are our assumptions/requirements on the data/outputs? Let's start there. Specifically, during loss calculation, i.e., `SIDLoss` and `WassersteinLoss`, the inputs (both true and predicted) be sum normalized? Similarly, when using these predictions at test time, must they also be sum normalized?

**Comment by am2145** [comment]
> So to be clear, what are our assumptions/requirements on the data/outputs? Let's start there. Specifically, during loss calculation, i.e., `SIDLoss` and `WassersteinLoss`, must the inputs (both true and predicted) be sum normalized? Similarly, when using these predictions at test time, must they also be sum normalized?

I believe the answer is yes to both based on v1's approach. In v1's `run_training.py` if `args.dataset_type` is "spectra", then the targets are sum normalized for the training, validation, and test sets. So I'd anticipate the trained model would predict (near) normalized spectra. In v1's `loss_functions.py`, both `sid_loss` and `wasserstein_loss` sum normalize the predictions and say the target_spectra must be normalized in their doc_strings. In v2, `SIDLoss` and `WassersteinLoss` handle the sum normalization of predictions during loss calculation, so I believe the only part we need to handle is the sum normalization of the targets in v2 to be consistent.

**Comment by davidegraff** [comment]
Ok so if I'm understanding correctly, during loss calculation both the target and predicted spectra must be sum-normalized. We have two options now:

1. To ensure correctness, we can perform this _inside_ the loss function's `__call__` (or `forward`--I forget which one it is) method. This guarantees by construction that the inputs will be valid for loss calculation. However, this is inefficient when the inputs might _already_ be sum-normalized and especially considering the same work will be repeated at each epoch.
2. We can _assume_ the inputs to the loss function sum-normalized (noting so in the docstring) and do this for the targets during dataset preparation. We will sum-normalize the predictions inside either `SpectralFFN.train_step` or `SpectralFFN.forward`, depending on the answer to the below question.

    **_NOTE_**: This approach does _not_ preclude sum-normalization inside the loss function. In this approach, we can also consider adding a `check_inputs: bool` parameter to the `__init__` method of the loss function which, if `True`, will perform sum-normalization during loss calculation and otherwise do nothing.

---

**The open question I have relates to test-time**: should the predicted spectra must be sum normalized as well, i.e., are the unnormalized outputs useful at all? If not, then we should _always_ sum-normalize inside (i.e., inside `SpectralFFN.forward`). Otherwise, we should only do so during prediction (i.e., inside `SpectralFFN.train_step`)

**Comment by am2145** [comment]
I think option 2 works well and can implement the `check_inputs: bool` for `SIDLoss()` and `WassersteinLoss()` so that we can also normalize the targets during loss calculation if it is desired.

As for whether the predicted spectra need to be sum-normalized, this is a really good question that I discussed briefly with the wider team today. There could be applications where predicting a non-normalized spectra is desired, but so far the previous application in v1 sum-normalize the target spectra during dataset preparation and so the model would be trained to predict normalized spectra as well.  Given that the targets have sum-normalization enfored, I think I'd do the same for the predictions.

To un-normalize the spectra in a physical sense, I guess we could save the sum of the inputs, and apply this to unscale the target and predicted spectra at test_time so that the peaks now have absorbance values that reflect the experimentally measured/computed values that were given as inputs. However, when it comes to using the model for inference on a new species, we wouldn't have the true spectra absorbance values to do the unscaling. For these cases, I think it is beneficial for the model predictions to return a sum-normalized spectra, as it makes it easier to compare different spectra consistently. 

So to summarize, I would have the predicted spectra be sum-normalized at test-time, but I will read up more and am of course happy to hear more thoughts on this.

**Comment by davidegraff** [comment]
> As for whether the predicted spectra need to be sum-normalized, this is a really good question that I discussed briefly with the wider team today. There could be applications where predicting a non-normalized spectra is desired, but so far the previous application in v1 sum-normalize the target spectra during dataset preparation and so the model would be trained to predict normalized spectra as well. Given that the targets have sum-normalization enfored, I think I'd do the same for the predictions.

To be clear, the model is **not** trained to predict a sum-normalized spectra. Rather, it is trained to minimize the SID or 1-Wasserstein distance between predicted and target spectra $P$ and $Q$, respectively. Both of these losses are metrics of distribution difference and therefore take normalized distributions as input. If we wanted our model to predict a normalized distribution, we would probably need to add an MSE loss based on `Y_pred.sum(-1) - 1`. Given that it's trivial to normalize an unnormalized distribution `Y = Y / Y_pred.sum(-1)`, we don't need to do this. This is not to say that our model _can't_ learn to predict a normalized distribution, just that there's no reason for it to do so given that we guarantee normalization by construction (and don't penalize but on "non-normalization").

> To un-normalize the spectra in a physical sense, I guess we could save the sum of the inputs, and apply this to unscale the target and predicted spectra at test_time so that the peaks now have absorbance values that reflect the experimentally measured/computed values that were given as inputs. However, when it comes to using the model for inference on a new species, we wouldn't have the true spectra absorbance values to do the unscaling. For these cases, I think it is beneficial for the model predictions to return a sum-normalized spectra, as it makes it easier to compare different spectra consistently.

Again, we need to be clear on what's necessary and simply convenient. We can't recover the unnormalized distribution of a prediction without storing the normalizing constant somewhere. The model is trained to minimize the "difference in shape" between the two spectra, so an unnormalized prediction can be interpreted as the raw absorbance values. That said, there's no penalty added for difference in normalizing constants ("unnormalized densities,") so the raw prediction and target are generally not directly comparable. If we think there will be some reason for users to utilize the raw predictions though, then we should not be silently normalizing for them.

**Comment by am2145** [comment]
> To be clear, the model is **not** trained to predict a sum-normalized spectra. Rather, it is trained to minimize the SID or 1-Wasserstein distance between predicted and target spectra P and Q, respectively. Both of these losses are metrics of distribution difference and therefore take normalized distributions as input. If we wanted our model to predict a normalized distribution, we would probably need to add an MSE loss based on `Y_pred.sum(-1) - 1`. Given that it's trivial to normalize an unnormalized distribution `Y = Y / Y_pred.sum(-1)`, we don't need to do this. This is not to say that our model _can't_ learn to predict a normalized distribution, just that there's no reason for it to do so given that we guarantee normalization by construction (and don't penalize but on "non-normalization").

I apologize for not being more precise. I know that the model does not predict sum-normalized spectra. What I meant was,  if the target spectra is sum-normalized and the loss function is used to minimize the distance to this sum-normalized target spectra, the predicted spectra's absorbances tend to be very close to sum-normalized values when using the spectralFFN(). However, I get your point that that's not the same as making it predict a normalized distribution or normalizing the predictions in the forward / training steps ourselves, so I should have been more clear about this.    

> Again, we need to be clear on what's necessary and simply convenient. We can't recover the unnormalized distribution of a prediction without storing the normalizing constant somewhere. The model is trained to minimize the "difference in shape" between the two spectra, so an unnormalized prediction can be interpreted as the raw absorbance values. That said, there's no penalty added for difference in normalizing constants ("unnormalized densities,") so the raw prediction and target are generally not directly comparable. If we think there will be some reason for users to utilize the raw predictions though, then we should not be silently normalizing for them

I guess what I'm uncertain about is the inference cases. During traing/val/test, we can train the model to the normalized target spectra and unscale its using the normalizing constant from the target data to get a spectra on the scale of the raw target absorbance values. If we apply this trained model to predict spectra for new species, our predictions would be on the scale of the normalized targets it was trained to, and the normalizing constant would not be known. In this case, if we have predictions for new species A and B, I'm wondering how we would best compare them. From our metrics and the literature, normalizing spectra for comparing/computing similarity is standard, as raw absorbances likely vary based on experiment/machine/setup/computational method etc. So to compare A and B, a potential workflow would be to normalize the model's predictions and then compute SID, SIS, Wasserstein, etc. Given this, I do think that it's beneficial to have normalized predicted spectra to facilitate comparison for these inference cases, but I acknowledge it is not necessary, so I'm not sure what the best approach is. 

I apologize for the long text dumps, but this discussion is very helpful. Thanks again David.

**Comment by davidegraff** [comment]
No worries. If I understand your response correctly, it sounds like should always be normalizing our predicted spectra. That tracks with the experimental procedure as well. That is, you wouldn't directly compare two spectra from different runs. Rather, you'd normalize their densities first and then compare. This makes me more confident in the decision to restructure the losses to assume normalized inputs (with a default `check_inputs: bool = False`).

**To summarize**:
- During dataset preparation, we normalize the densities of target spectra
- During spectral prediction (i.e., in `forward`,) we normalize the densities of the predicted spectra
- During loss calculation, we assume both inputs are normalized _only if_ `not self.check_inputs`. By default, this will be `True` (because we want to avoid repeating work).

**Comment by am2145** [comment]
Sounds good to me. I believe the small change I have in this PR handles item 2 on the bullet point list. As for items 1 and 3, I'm happy to implement those, presumably in separate PRs? Or would it be preferred to have them as part of this one? In either case, if there are any edits to make/changes before opening up for review, I'm happy to make those too.

**Comment by davidegraff** [comment]
Fair point, I think (2) is in-scope of this PR while (1) and (3) should be separate.

**Comment by am2145** [comment]
Got it. I'll work on separate PRs for (1) and (3). I have opened up the PR for review so that we can address (2).

**Review by davidegraff — COMMENTED** [review]
didn't we agree on all outputs of a spectral predictor being sum-normalized?

**Comment by am2145** [comment]
Yes, sorry about that. I'll add the sum-normalization to the training step after resolving some conflicts on my local copy and push that here.

File: chemprop/nn/predictors.py [hunk 1]
```diff
@@ -297,3 +297,14 @@ def __init__(self, *args, spectral_activation: str | None = "softplus", **kwargs
...
+
+    def forward(self, Z: Tensor) -> Tensor:
+        Y = super().forward(Z)
+        Y = self.ffn.spectral_activation(Y)
+        return Y / Y.sum(1, keepdim=True)
+
+    def train_step(self, Z: Tensor) -> Tensor:
+        Y = super().forward(Z)
+        Y = self.ffn.spectral_activation(Y)
+        return Y / Y.sum(1, keepdim=True)
```
Comment (davidegraff) [inline chemprop/nn/predictors.py:309]: It's best to avoid copying code between methods:
```suggestion
   train_step = forward
```
This assigns the `train_step` method to the `forward` method. It looks like attribute assignment because, mechanically, it's the same thing in python.

Comment (am2145) [reply chemprop/nn/predictors.py]: I see. Thanks for explaining. I have committed the suggestion.

**Review by davidegraff — APPROVED** [review]

**Comment by KnathanM** [comment]
> Got it. I'll work on separate PRs for (1) and (3). I have opened up the PR for review so that we can address (2).

While you work on those PRs, could you open an issue for them to help us track their completion. Thank you for your work on this.
